docs/snapshotters/blockfile.md

```diff
@@ -0,0 +1,145 @@
+
+### Creating the scratch file
+
+You can create a scratch file as follows. This example uses a 500MB scratch file.
+
+```bash
+$ # make a 500M file
+$ dd if=/dev/zero of=/opt/containerd/blockfile bs=1M count=500
+500+0 records in
+500+0 records out
+524288000 bytes (524 MB, 500 MiB) copied, 1.76253 s, 297 MB/s
+
+$ # format the file with ext4
+$ sudo mkfs.ext4 /opt/containerd/blockfile
+mke2fs 1.47.0 (5-Feb-2023)
+Discarding device blocks: done
+Creating filesystem with 512000 1k blocks and 128016 inodes
+Filesystem UUID: d9947ecc-722d-4627-9cf9-fa2a3b622106
+Superblock backups stored on blocks:
+        8193, 24577, 40961, 57345, 73729, 204801, 221185, 401409
+
+Allocating group tables: done
+Writing inode tables: done
+Creating journal (8192 blocks): done
+Writing superblocks and filesystem accounting information: done
+```
+
+### Running a container
+
+To run a container using the blockfile snapshotter, you need to specify the
+snapshotter:
+
+```bash
+$ # ensure that the image we are using exists; it is a regular OCI image
+$ ctr image pull docker.io/library/busybox:latest
+$ # run the container with the provides snapshotter
+$ ctr run -rm -t --snapshotter blockfile docker.io/library/busybox:latest hello sh
+```
+
+To use it via the go client API, it is identical to using any other snapshotter:
+
+```go
+// TODO: will fill in
+```
+
+## How It Works
+
+The blockfile snapshotter functions similarly to other snapshotters. It uses the
+provided image to find layers, and then applies each on top of the previous,
+creating a usable directory.
+
+The blockfile snapshotter is unique in two ways:
+
+1. It applies layers inside a disk image file, rather than on the host filesystem.
+1. It creates a block image file for each layer, applying the previous on top of it.
+
+For every layer the snapshotter creates a new blockfile, starting with a copy of the
+blockfile from the previous layer. If there is no previous layer, i.e. for the first
+layer, it copies the scratch file.
+
+For example, for an image with 3 layers - called A, B, C - the process is as follows:
+
+1. Copy the scratch file to a new blockfile for layer A.
+1. Loopback-mount the blockfile for layer A.
+1. Apply layer A to the mount.
+1. Unmount the blockfile for layer A.
+1. Copy the blockfile for layer A to a new blockfile for layer B.
+1. Loopback-mount the blockfile for layer B.
+1. Apply layer B to the mount.
+1. Unmount the blockfile for layer B.
+1. Copy the blockfile for layer B to a new blockfile for layer C.
+1. Loopback-mount the blockfile for layer C.
+1. Apply layer C to the mount.
+1. Unmount the blockfile for layer C.
+
+Each unpack of a layer builds upon the contents of the previous layers into a new
+blockfile.
+
+TODO: Does this mean a lot of wasted space? Does it remove previous layer blockfiles after they have been used? E.g. a 500MB scratch image copied for layer A, then for B, then for C, means 1.5GB of space used, even though the final disk image is only 500MB.
```

Comment:
We try and use every OS' (assuming it's supported by the underlying filesystem also) sparse file copy support. Even if the blockfile is 500MB, if there's only 25MB of layer content inside, the copy of the blockfile to the next snapshot will only actually copy 25MB. It may show up in some unix tools as using 500MB but it won't actually be.

Reply:
So we keep the interim blockfiles around, but because they are sparse, it is not quite that inefficient.

Still, if the scratch is 500MB, and each of 3 layers is 50MB, then I have:

* layer A - 50MB
* layer B - 100MB (50 from A plus 50 from B)
* layer C - 150MB (50 + 50 + 50)

Total space isn't 150MB, but 300MB.

Reply:
Yes the above is correct